### PR TITLE
{Audio,Video}Stream: use AsyncIterableIterator

### PR DESCRIPTION
**Pull request by nbsp**
instead of having a typed EventEmitter

### CODE REVIEW

File: packages/livekit-rtc/src/audio_stream.ts [hunk 1]
```diff
@@ -70,15 +59,36 @@ export class AudioStream extends (EventEmitter as new () => TypedEmitter<AudioSt
...
+  async next(): Promise<IteratorResult<AudioFrame>> {
+    if (this.eventQueue.length > 0) {
+      const value = this.eventQueue.shift();
+      if (value) {
+        return { done: false, value };
+      } else {
+        return { done: true, value: undefined };
+      }
+    }
+    return new Promise((resolve) => (this.queueResolve = resolve));
```
Comment (lukasIO) [inline packages/livekit-rtc/src/audio_stream.ts:83]: hm. I think `this.queueResolve` might get overridden before being called?
so the async iterator could get stuck

Comment (lukasIO) [reply packages/livekit-rtc/src/audio_stream.ts:83]: the `Mutex` helper that Théo has mentioned already might make sense here

Comment (nbsp) [reply packages/livekit-rtc/src/audio_stream.ts:83]: could it? i don't think that's the case, it only gets overridden if you call `next()` twice without awaiting the first one, which you shouldn't do anyway.

Comment (lukasIO) [reply packages/livekit-rtc/src/audio_stream.ts:83]: > which you shouldn't do anyway.

agreed that it _shouldn't_ be done, but seems like a footgun regardless?

Comment (nbsp) [reply packages/livekit-rtc/src/audio_stream.ts:83]: would it make sense to return a copy of the same promise if `next` is called twice?

Comment (lukasIO) [reply packages/livekit-rtc/src/audio_stream.ts:83]: I'm not sure, I think best case scenario is a promise chain, which is exactly what the `Mutex` class does.

**Review by theomonnom — APPROVED** [review]

**Comment by theomonnom** [comment]
Can you do the same for VideoStream?

**Pull request merged**